EditorRegions: Move to block-editor package, rename to __experimentalEditorSkeleton

ockham opened the pull request:
## Description

Move the `EditorRegions` component from `packages/post-editor` to `packages/block-editor`, and rename to `__experimentalEditorSkeleton`. This component is generic enough that it might come in handy for other consumers -- e.g. `packages/edit-site`, and 3rd parties. (The specifically intended use case will be a 'custom' build of Gutenberg, with some added UI controls etc.) If this is approved, I hope to continue to extract other reusable components similarly.

## How has this been tested?

Rebuild Gutenberg (`npm run build`), and start it (`npx wp-env start`). Verify that the editor styling isn't broken (specifically regions like header or sidebar).

## Types of changes
Move a component from one package to another. Update style selectors, but retain old class names along new ones for backwards compat.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

ockham committed:
EditorRegions: Move to editor package

ockham committed:
Move style import

ockham committed:
Update EditorRegions import

ockham committed:
Export EditorRegions from editor package

ockham committed:
s/edit-post-editor-regions/components-editor-regions/g

ockham committed:
Add new class names to EditorRegions component

gziolo commented:
@youknowriad is your friend, he has a better idea of why this component is in `@wordpress/edit-post`. I think this component is quite opinionated and I can't tell if it's applicable for `edit-site` or `edit-widgets`. However, I understand the point that other projects would like to reuse it to keep the same structure of the content. The only challenge is that kind of components might change the order of items if there is a need to improve flows.

youknowriad commented on packages/editor/src/components/editor-regions/index.js:
```diff
@@ -18,10 +18,16 @@ function EditorRegions( {
+		<div
+			className={ classnames(
+				className,
+				'components-editor-regions',
```
The convention is not right, it should be `editor-editor-regions` (same for the sub classes).
Also wondering if there's a better name than EditorRegions

ockham replied:
> The convention is not right, it should be `editor-editor-regions` (same for the sub classes).

Yeah, was wondering about the class prefix. FWIW, I went followed precedent in `packages/editor/src/components/`: https://github.com/WordPress/gutenberg/blob/164d5830a9acd895dfd661b5fde1ca8b80b270ac/packages/editor/src/components/editor-notices/index.js#L35

> Also wondering if there's a better name than EditorRegions

:thinking: Any suggestions? Maybe... EditorScaffolding? EditorSkeleton? EditorChrome?

youknowriad replied:
I think I like both EditorSkeleton and EditorChrome. Maybe more Skeleton? thoughts?
And yeah that notices class name is wrong too.

ockham replied:
Yeah, Chrome could be misleading. I hope 'Skeleton' won't be off-putting for people? :sweat_smile: :skull:  EditorStructure maybe?

ockham replied:
I went with `EditorSkeleton` :slightly_smiling_face:

youknowriad commented on packages/editor/src/components/editor-regions/index.js:
```diff
@@ -18,10 +18,16 @@ function EditorRegions( {
+					className="components-editor-regions__header edit-post-editor-regions__header"
 					role="region"
 					/* translators: accessibility text for the top bar landmark region. */
 					aria-label={ __( 'Editor top bar' ) }
```
I was wondering whether the aria-label are too opinionated and should be made customizable or not. Potentially, they couuld work for any "editor".

ockham replied:
Yeah, maybe :thinking: I'd go with YAGNI here -- don't make it customizable until needed?

youknowriad commented on packages/editor/src/components/editor-regions/index.js:
```diff
@@ -18,10 +18,16 @@ function EditorRegions( {
 	className,
```
Why did you choose the "editor" package. The editor package is about the "post editor" (tied to a WP Post object). 
I was thinking this component could be useful in `edit-widgets` too which doesn't depend on the "editor" package.

In fact, I feel we should try to refactor `edit-widgets` to use it. It could be a good proof that this component is generic enough to be shared.

youknowriad replied:
Potentially, it could be in "block-editor" (we can try this but make it experimental as a start). I feel this is more something like: `@wordpress/admin-ui-base` or something like that but maybe we should avoid a new package for just one component for now.

ockham replied:
> Why did you choose the "editor" package. The editor package is about the "post editor" (tied to a WP Post object).

I thought that's what `post-edit` was about :sweat_smile: But yeah, I wasn't totally sure where to put it. `packages/components/` didn't seem right (that's more for UI controls).

> I was thinking this component could be useful in `edit-widgets` too which doesn't depend on the "editor" package.

Ah yeah, good point :+1: 

> In fact, I feel we should try to refactor `edit-widgets` to use it. It could be a good proof that this component is generic enough to be shared.

As part of this PR? :thinking: Or fine to do it in follow-up?

> Potentially, it could be in "block-editor" (we can try this but make it experimental as a start). I feel this is more something like: `@wordpress/admin-ui-base` or something like that but maybe we should avoid a new package for just one component for now.

Yeah, my thoughts exactly. I was originally moving it into a newly created `editor-scaffolding` package, but decided against it. Anyway, I like the idea of putting it into `block-editor` for starters, and maybe to a newly created package once the need arises.

ockham committed:
Move to block-editor

ockham committed:
Fix exports/imports

ockham committed:
s/components-editor-regions/block-editor-editor-regions/g

ockham committed:
Change classnames to block-editor-editor-skeleton*

ockham committed:
Rename to EditorSkeleton

ockham committed:
Prefix as __experimental

ockham commented:
I moved the component to the `block-editor` package, and renamed it to `__experimentalEditorSkeleton`. @youknowriad Care to take another look? :blush:

youknowriad commented on packages/block-editor/src/components/editor-skeleton/index.js:
```diff
@@ -18,10 +18,16 @@ function EditorRegions( {
+			className={ classnames(
+				className,
+				'block-editor-editor-skeleton',
+				'edit-post-editor-regions'
```
I think we can drop this className tbh. Did you leave it for BC? This seems to be ready to land before beta1 so there shouldn't be any problem to remove the classname.

ockham replied:
Yeah, I left it in for backwards compat. I know of at least [one instance in Calypso](https://github.com/Automattic/wp-calypso/blob/0fe9f9d07ba498db6b1bc89fc3aeb767a84be392/apps/wpcom-block-editor/src/calypso/features/iframe-bridge-server.scss#L19) where we're referring to this class name, and since we were having quite some trouble with class name changes in the past (specifically with the `.editor-` -> `.block-editor-` change, see https://github.com/Automattic/wp-calypso/pull/38857, #14420, and #19050), I wanted at least to give a grace period before dropping the old class name.

(Then again, a point can be made that this was never part of a public interface, and that that grace period didn't help in updating the affected Calypso selectors in time before they were dropped in the past either, so :man_shrugging: )

youknowriad commented:
> As part of this PR? 🤔 Or fine to do it in follow-up?

Whatever you think is best, that said, I think it's important to validate that this component is generic enough to be shared properly.

youknowriad reviewed (approved):
Thanks 👍

ockham commented:
> > As part of this PR? thinking Or fine to do it in follow-up?
> 
> Whatever you think is best, that said, I think it's important to validate that this component is generic enough to be shared properly.

I'll file a follow-up; if there are any changes needed to `EditorSkeleton`, I'll make them there.

youknowriad commented:
you can merge since it's approved. I figured you might waiting for me?

ockham commented:
> you can merge since it's approved. I figured you might be waiting for me?

Thanks for merging! I was considering dropping [the BC stuff we discussed here](https://github.com/WordPress/gutenberg/pull/20050#discussion_r375887591) as part of this PR but didn't get to that on Friday :sweat_smile:
